librad/src/git/replication.rs

```diff
@@ -181,119 +224,248 @@ where
+mod person {
+    use super::*;
+
+    #[allow(clippy::unit_arg)]
+    #[tracing::instrument(level = "trace", skip(storage), err)]
+    pub fn ensure_setup(
+        storage: &Storage,
+        person: Person,
+        remote_peer: PeerId,
+    ) -> Result<(), Error> {
+        let urn: Urn = Reference::rad_id(Namespace::from(person.urn()))
+            .with_remote(remote_peer)
+            .try_into()
+            .expect("namespace is set");
 
-            // Track all delegations
-            for key in person.into_inner().doc.delegations {
-                tracking::track(storage, &urn, PeerId::from(key))?;
-            }
+        match identities::person::verify(storage, &urn)? {
+            None => Err(Error::MissingIdentity),
+            Some(person) => {
+                // Create `rad/id` here, if not exists
+                ensure_rad_id(storage, &urn, person.content_id)?;
 
-            Ok(())
-        },
+                // Track all delegations
+                for key in person.into_inner().doc.delegations {
+                    tracking::track(storage, &urn, PeerId::from(key))?;
+                }
+
+                Ok(())
+            },
+        }
     }
 }
 
-#[tracing::instrument(level = "trace", skip(storage), err)]
-fn ensure_setup_as_project(
-    storage: &Storage,
-    proj: Project,
-    remote_peer: PeerId,
-) -> Result<impl Iterator<Item = Urn>, Error> {
-    let local_peer_id = storage.peer_id();
+mod project {
+    use super::*;
 
-    let urn: Urn = Reference::rad_id(Namespace::from(proj.urn()))
-        .with_remote(remote_peer)
-        .try_into()
-        .expect("namespace is set");
+    #[derive(Clone, Debug)]
+    pub struct DelegateProject {
+        pub urn: Urn,
+        pub project: VerifiedProject,
+    }
 
-    // Verify + symref the delegates first
-    for delegate in proj.delegations().iter().indirect() {
-        let delegate_urn = delegate.urn();
-        // Find in `refs/namespaces/<urn>/refs/remotes/<remote
-        // peer>/rad/ids/<delegate.urn>`
-        let in_rad_ids: Urn = Reference::rad_delegate(Namespace::from(&urn), &delegate_urn)
+    pub fn verify(storage: &Storage, urn: Urn, remote: PeerId) -> Result<VerifiedProject, Error> {
+        let urn: Urn = Reference::rad_id(Namespace::from(urn.clone()))
+            .with_remote(remote)
+            .try_into()
+            .expect("namespace is set");
+
+        Ok(identities::project::verify(storage, &urn)?.ok_or(Error::MissingIdentity)?)
+    }
+
+    pub fn verify_delegates(
+        storage: &Storage,
+        urn: Urn,
+        delegates: BTreeMap<PeerId, Urn>,
+    ) -> Result<BTreeMap<PeerId, DelegateProject>, Error> {
+        let mut confirmed_delegates = BTreeMap::new();
+        for remote in delegates.keys() {
+            match verify(&storage, urn.clone(), *remote) {
+                // It's possible that this PeerId for the delegate didn't live on the machine we
+                // fetched from
```

Comment:
But that's a protocol violation, and should abort the fetch

Reply:
Sorry, this isn't clear. The issue here is that a `Person` can have multiple `Direct` delegates. Only _one_ of these may be the remote `PeerId` on the machine we're replicating from. 

So the invariant is that there must be at least one `PeerId` for `Urn` but we need to find out which one. I suppose we could check that when building the map?